File: src-server/src/db/schemas/provider.ts
import { integer, sqliteTable, text } from "drizzle-orm/sqlite-core";
import { z } from "zod";

const llmModelSchema = z.object({
  id: z.string(),
  contextWindow: z.int(),
  capabilities: z.object({
    toolUse: z.boolean().optional(),
    vision: z.boolean().optional(),
  }),
});

const llmProviderType = z.enum(["openai", "anthropic", "gemini"]);

export type LlmModel = z.infer<typeof llmModelSchema>;
export type LlmProviderType = z.infer<typeof llmProviderType>;

export const providers = sqliteTable("providers", {
  id: integer("id").primaryKey({ autoIncrement: true }).unique(),
  name: text("name").notNull(),
  type: text("type").$type<LlmProviderType>().notNull(),
  apiKey: text("api_key").notNull(),
  baseUrl: text("base_url"),
  models: text("models", { mode: "json" }).$type<LlmModel[]>().notNull(),
});

export type Provider = typeof providers.$inferSelect;

File: src-server/db/schemas/provider.py
from pydantic import BaseModel, ConfigDict
from ..models.provider import ProviderType, LlmModelCapability

class LlmModel(BaseModel):
    name: str
    context_size: int
    capability: LlmModelCapability

class ProviderBase(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    name: str
    type: ProviderType
    base_url: str
    api_key: str
    models: list[LlmModel]

class ProviderRead(ProviderBase):
    id: int

class ProviderCreate(ProviderBase):
    pass

class ProviderUpdate(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    name: str | None = None
    type: ProviderType | None = None
    base_url: str | None = None
    api_key: str | None = None
    models: list[LlmModel] | None = None
